Improve formatHost

nyamsprod opened the pull request:
- simplify host formatting
- improve IDN to ASCII conversion
- handles invalid host properly

@kelunik this should be in line of what you are suggesting. The errors reason/phrase could be improved. Feedback appreciated.

nyamsprod committed:
Improve formatHost

- simplify host formatting
- improve IDN to ASCII conversion
- handles invalid host properly

kelunik commented on src/AbstractUri.php:
```diff
@@ -374,21 +374,52 @@ protected function formatHost($host)
+        $component = strtolower($host);
+        if (false !== strpos($component, '%')) {
+            $component = rawurldecode($component);
+        }
```
I guess this could be moved above the `preg_match` check, but I'm not sure. Actually I wasn't aware host names could be encoded, where is this specified?

nyamsprod replied:
it is written in [RFC3986](https://tools.ietf.org/html/rfc3986#section-3.5)

> reg-name    = *( unreserved / pct-encoded / sub-delims )
where pct-encoded is
>  pct-encoded = "%" HEXDIG HEXDIG

kelunik replied:
Wrong section, but indeed, TIL. https://tools.ietf.org/html/rfc3986#section-3.2.2 is the right one. Most domains won't be encoded, especially if they're ASCII only, so let's keep the check below the regex. This is potentially slower if an ASCII only domain is encoded, but it makes the much more common case of no encoding and an ASCII only domain faster.

kelunik replied:
This `if` should be moved below the `$reg_name` `preg_match`, because that will favor the happy path.

kelunik commented on src/AbstractUri.php:
```diff
@@ -374,21 +374,52 @@ protected function formatHost($host)
+            return $component;
+        }
+
+        throw new Exception(sprintf('Host %s can not be converted : %s', $host, $this->getIDNErrors($arr['errors'])));
```
I think this should be `'Host %s is invalid: %s'`

kelunik commented on src/AbstractUri.php:
```diff
@@ -374,21 +374,52 @@ protected function formatHost($host)
+     *
+     * @return string
+     */
+    protected function getIDNErrors(int $error_byte): string
```
Any reason for this to be protected instead of private?

nyamsprod commented:
After a good night sleep ... I think these improved checks should be added is the Uri\is_host function instead since this is were host value is validated before being formattedd by the formatHost method. I'll look into adding them there instead. the formatHost method as the name implied should only care about formatting the host.

kelunik commented:
I'd rather drop the `is_host` check in `createFromComponents` and `withHost` and check it in `formatHost`, otherwise it's always doing double the work, and `idn_to_ascii` is quite costly.

nyamsprod commented:
@kelunik Agreed we need to reduce `idn_to_ascii` calls. I'll try to update the PR then and see the result of removing the call to is_host

nyamsprod committed:
Improve IDNA error reporting

nyamsprod committed:
getIDNErrors is private static

nyamsprod committed:
Improve Factory::create method

- URI resolve should be apply on any given URI not only
when a base URI is given

nyamsprod committed:
Revert "Improve Factory::create method"

This reverts commit b89527148230003c60aa460fadb4ae1137cbffce.

nyamsprod committed:
Improve AbstractUri::formatHost

nyamsprod commented:
@kelunik All the checks are now done in `formatHost` . This could be further improve IMHO once the PR on uri parser is merged.... thoughts ? If all is good for you I will merge this.

nyamsprod committed:
Improve filterPort and __toString

nyamsprod committed:
improve type hinting

nyamsprod committed:
Improve internal code

kelunik commented on src/AbstractUri.php:
```diff
@@ -362,33 +356,76 @@ protected static function urlEncodeMatch(array $matches): string
      * @return string|null
      */
-    protected function formatHost($host)
+    protected function formatHost(string $host = null)
```
This is a BC break unless you require PHP 7.2.

nyamsprod replied:
I'll revert this change

kelunik commented on src/AbstractUri.php:
```diff
@@ -362,33 +356,76 @@ protected static function urlEncodeMatch(array $matches): string
             return $host;
         }
 
+        static $pattern = '/^(?<name>[a-z0-9]([a-z0-9-]{0,61}[a-z0-9])?)(?:\.(?&name))*$/i';
```
As we learned, hosts don't need to comply to DNS rules, therefore the limit of label lengths is unnecessary. This should just be a regex that checks for `reg-name` I think.

nyamsprod replied:
indeed but adding this check improve the performance for the majority of cases. Most host do use the DNS system. So it's a optimization for the common usage.

kelunik replied:
Doesn't the below `is_host` catch those?

kelunik commented on src/AbstractUri.php:
```diff
@@ -362,33 +356,76 @@ protected static function urlEncodeMatch(array $matches): string
-            if (strlen($label) !== strspn($label, $valid_ascii_label_characters)) {
-                $label = (string) idn_to_ascii($label, IDNA_NONTRANSITIONAL_TO_ASCII, INTL_IDNA_VARIANT_UTS46);
-            }
+        if (false !== strpos($host, ']')) {
```
Any reason why this isn't `($host[0] ?? '') === '['`? Saves the function call and iterating the whole string internally.

nyamsprod replied:
none I'll use this check instead

kelunik commented on src/AbstractUri.php:
```diff
@@ -398,7 +435,7 @@ protected function formatHost($host)
      * @return int|null
      */
-    protected function formatPort($port)
+    protected function formatPort(int $port = null)
```
Same here, potential BC break.

nyamsprod replied:
I'll revert this changes ..

kelunik commented:
@nyamsprod I added a few more comments. Do you have any plans for a new major version? Given the changes to parser might be rather significant, should that be considered to go into a new major?

nyamsprod commented:
@kelunik yes I'm thinking about this too I already have some idea the main one being to better organize the library. Right now there are too many repos (6). I believe 3 or 4 are enough.

- 1 for the parsers (Uri/Query) 
- 1 for the schemes
- 1 for the components
- (and maybe one which will contains only the interfaces - or this one can be added in the parser repo ?)

Since all packages depends on the parser if we improve the parser drastically then yeah basically it means a new major should be in the work. I'll try to create a project on the main repo to better explain these changes. 

Independently, adding some performance boost in v5 is still a good idea.

kelunik commented:
@nyamsprod The main meta package doesn't make too much sense IMO, because any breaking change in any of the packages is a breaking in the meta package.

We'd definitely like `components` to stay separate, as we don't like to have an `ext-curl` dependency for a URI library. The `0.8.x` development branch of Aerys only depends on `parser` and `schemes` for that reason.

IMO it makes sense to keep the query parsing separate. The interpretation of the query varies significantly between languages. PHP's `parse_str` is nowhere standard. In fact, I think the current implementation of query parsing is sub-optimal. You'll have to check whether something is a string or array of strings every time. I'd be better to offer something like `getParam()` and `getParamArray()`, where the first always returns the first instance and the second always returns an array, even if there's just one or even none. Also, `getParam()` and `getPair()` or whatever it is called is highly confusing TBH.

I'll just leave this here, because it's already written. Let's wait with any further discussion until an appropriate issue exists instead of having it in a PR.

nyamsprod commented:
@kelunik `getParam` is for getting parameter `à la PHP` while `getPair` returns the parameters as key/pair assoc. like [UrlSearchParam](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams/get) in Javascript. But yeah maybe the naming can be improved.

`ext-curl` is not really required if it is then it's for default usage... It could be removed from the required section. It is a left over from using PHP Domain parser. this kind of requirement could be removed in a ne version.

Last but not least indeed, the Uri meta package could/can be scrap it is also in my mind 👍

kelunik commented:
@nyamsprod I know, but I had to dig into the code for that, it wasn't clear from the documentation. Should have probably opened an issue for that.

`ext-curl` isn't a strict requirement, but it's used in that code, so … anyway, we don't need any of the functionality there, it's fine the way it is.

nyamsprod committed:
revert possible BC Break

nyamsprod committed:
Improve Host formatting

kelunik commented on src/AbstractUri.php:
```diff
@@ -362,30 +362,96 @@ protected function formatHost($host)
+        $res = array_reduce(str_split(unpack('A16', inet_pton($ipv6))[1]), $reducer, '');
+        if (substr($res, 0, 10) === $local_link_prefix) {
             return $host;
         }
```
This code looks highly complicated, could you add a comment what exactly it does?

nyamsprod replied:
updated with comments

kelunik replied:
How about simply https://3v4l.org/lMRh5?

```php
var_dump(substr(inet_pton("fe80::1") & "\xfe\x80", 0, 2) === "\xfe\x80");
var_dump(substr(inet_pton("fe3a::1") & "\xfe\x80", 0, 2) === "\xfe\x80");
```

nyamsprod replied:
I can update the code if your take is simpler. I use this code because I understood it well but yes yours is easier so I'll update this part and see if the test suite is ok with it 👍

nyamsprod committed:
adding comments to explain zone Id detection

nyamsprod committed:
let's favor the happy path

nyamsprod committed:
improve AbstractUri::formatScheme

nyamsprod committed:
improve local link detection

nyamsprod committed:
Adding IPvfuture support

nyamsprod committed:
improve IP Future detection

nyamsprod committed:
Improve Port filtering

kelunik commented on src/AbstractUri.php:
```diff
@@ -384,25 +384,28 @@ private function formatIp(string $host)
-            v(?<version>\d+)\.
-            ((?<unreserved>[a-z0-9_~\-\.])|(?<sub_delims>[!$&\'()*+,;=:]))+
+            v(?<version>[A-F0-9])+\.
+            (
```
This should be a non-matching group, so `(?:` instead of `(`.

nyamsprod replied:
good catch will change that

nyamsprod committed:
using non-matching group for ipfuture

nyamsprod committed:
Improve benchmark

The benchmark now focuses on the URI scheme code only
the parser should be benchmark on the URI parser repo
instead

kelunik commented on benchmark/parse.php:
```diff
@@ -1,7 +1,12 @@
+];
 for ($i = 0; $i < 100000; $i++) {
-    League\Uri\Http::createFromString('http://amphp.org/amp');
+    League\Uri\Http::createFromComponents($components);
```
It no longer benchmarks the parsing now, so the file should be renamed.

nyamsprod replied:
how would you call the file  ? simply `benchmark.php` ?

kelunik replied:
Maybe simply `http-create-from-components.php`?

kelunik commented on src/AbstractUri.php:
```diff
@@ -360,11 +360,11 @@ protected function formatHost($host)
 
-        if ('[' === $host[0] && ']' === substr($host, -1)) {
-            return $this->formatIp($host);
+        if ('[' !== $host[0] || ']' !== substr($host, -1)) {
```
There's no need for the `||` part, the first condition should be enough.

kelunik commented on src/AbstractUri.php:
```diff
@@ -448,15 +445,15 @@ private function formatRegisteredName(string $host)
         }
 
-        throw new UriException(sprintf('Host %s is invalid : %s', $host, $this->getIDNAErrors($arr['errors'])));
+        throw new UriException(sprintf('the submitted host `%s` is invalid : %s', $host, $this->getIDNAErrors($arr['errors'])));
```
IMO the previous exception message was better with just `Host '%s' is invalid: %s`.

Note: I don't like using ` there, as it doesn't work well with markdown. :P

nyamsprod replied:
I can revert that ... I don't mind at all

kelunik commented on src/AbstractUri.php:
```diff
@@ -297,13 +291,14 @@ public function __get(string $property)
-        if (is_scheme($scheme)) {
-            return $component;
+        $formatted_scheme = strtolower($scheme);
+        static $pattern = '/^[a-z][a-z\+\.\-]*$/i';
```
The `i` can be removed, as the string is already lowercase.

kelunik commented on src/AbstractUri.php:
```diff
@@ -353,42 +348,152 @@ protected static function urlEncodeMatch(array $matches): string
+            (?<unreserved>[a-z0-9_~\-\.])|
+            (?<sub_delims>[!$&\'()*+,;=])|
+            (?<encoded>%[A-F0-9]{2})
+        )+$/ix';
```
Again, no need for the `i` modifier here.

kelunik commented on src/AbstractUri.php:
```diff
@@ -473,9 +578,9 @@ protected function formatPath(string $path): string
      * @return string|null
      */
-    protected function formatQueryAndFragment($component)
+    protected function formatQueryAndFragment(string $component = null)
```
This needs to be reverted as well to not be a BC break.

kelunik commented on src/AbstractUri.php:
```diff
@@ -956,7 +1055,11 @@ public function withPort($port): self
+            return $port;
+        }
+
+        if ($port < 1) {
```
I don't think there's anything that says a port must be > 0.

nyamsprod replied:
no there's not ... but 

>  port        = *DIGIT

so maybe make `if ($port < 0)`

kelunik replied:
Yes, that's better. `0` is e.g. used by Linux in `tcp://127.0.0.1:0` to bind to a random available port.

nyamsprod committed:
Improve Port filtering

kelunik commented:
@nyamsprod Ready for squash and merge? :tada:

nyamsprod commented:
I like to leave it this way and come back the day after sometimes I stumble on stuffs I've missed 👍

nyamsprod commented:
so yes squash/merge tomorrow

kelunik commented:
@trowski Maybe you want to review, too?

nyamsprod committed:
Improve PR

trowski commented:
LGTM. :+1:

kelunik commented:
:tada:
